site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1489 @@ next: false
+  task: () => 1000n ** 1000n,
+  // Convert bigint to string for storage.
+  beforeStore: (n) => String(n),
+  // Convert string back to bigint for usage.
+  afterLoad: (str) => BigInt(str),
+});
 ```
 
-:::
+If you want to throw an error from the task, you can specify additional serialization functions for error objects.
+Check out [`ExternalOp`](/ref/conversations/externalop) in the API reference.
 
-Usually, outside of the conversations plugin, each of these updates would be handled by the [middleware system](../guide/middleware) of your bot.
-Hence, your bot would handle the update via a context object which gets passed to your handlers.
+### Storage Keys
 
-In conversations, you will obtain this new context object from the `wait` call.
-In turn, you can handle different updates differently based on this object.
-For example, you can check for text messages:
+By default, conversation data is stored per chat.
+This is identical to [how the session plugin works](./session#session-keys).
 
-::: code-group
+As a result, a conversation cannot handle updates from multiple chats.
+If this is desired, you can [define your own storage key function](/ref/conversations/conversationoptions#storage).
+As with sessions, it is [not recommended](./session#session-keys) to use this option in serverless environments.
 
-```ts [TypeScript]
-async function waitForText(conversation: MyConversation, ctx: MyContext) {
-  // Wait for the next update:
-  ctx = await conversation.wait();
-  // Check for text:
-  if (ctx.message?.text) {
-    // ...
-  }
-}
-```
+You can specify the storage key function as follows.
 
-```js [JavaScript]
-async function waitForText(conversation, ctx) {
-  // Wait for the next update:
-  ctx = await conversation.wait();
-  // Check for text:
-  if (ctx.message?.text) {
-    // ...
-  }
-}
+```ts
+bot.use(conversations({
+  storage: {
+    type: "key",
+    adapter: storageAdapter,
+    getStorageKey: (ctx) => ctx.from?.id.toString(),
+  },
+}));
 ```
 
-:::
+Check out the API reference for [`ConversationStorage`](/ref/conversations/conversationstorage) to see more details about storing data with the conversations plugin.
+Among other things, it will explain how to store data without a storage key function at all using `type: "context"`.
+
+## Using Plugins Inside Conversations
 
-In addition, there are a number of other methods alongside `wait` that let you wait for specific updates only.
-One example is `waitFor` which takes a [filter query](../guide/filter-queries) and then only waits for updates that match the provided query.
-This is especially powerful in combination with [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment):
+[Remember](#conversational-context-objects) that the context objects inside conversations are independent from the context objects in the surrounding middleware.
+This means that they will have to plugins installed on them by default.
```

Comment:
Not sure, but shouldn’t this be just _will have plugins_?

Reply:
```suggestion
This means that they will have no plugins installed on them by default.
```